### PR TITLE
rspconfig in python to get network information

### PR DESCRIPTION
**Pull request by zet809**
For task #4674 

`Dependency needed:`

* For rpm package:                     ||  For deb:
  * gcc                                          ||
  * libffi-devel                              ||    * libffi-dev
  * python-devel                          ||    * python-dev
  * openssl-devel                         ||    * libssh2-1-dev
* For pip package:
  * paramiko
  * cryptography
  * ssh2-python==0.8.0
  * parallel-ssh
  * scp

[root@briggs01 ~]# XCAT_OPENBMC_PYTHON=YES rspconfig mid05tor12cn02 ip netmask gateway ipsrc vlan
Processing command rspconfig in openbmc2.pm
mid05tor12cn02: Found LinkLocal address 169.254.164.2 for interface /xyz/openbmc_project/network/eth0, Ignoring...
mid05tor12cn02: Found LinkLocal address 169.254.20.180 for interface /xyz/openbmc_project/network/eth0_11, Ignoring...
Netinfo: {'hostname': u'mid05tor12cn02-bmc-test', 'defaultgateway': u'172.11.253.27', u'eth0_11': {'mac': u'70:e2:84:14:28:97', 'ntpservers': [], 'netmask': 16, 'ip': u'172.11.139.2', 'ipsrc': u'Static', 'gateway': u'0.0.0.0', 'vlanid': 11}}
dic_length: 1: {u'eth0_11': {'mac': u'70:e2:84:14:28:97', 'ntpservers': [], 'netmask': 16, 'ip': u'172.11.139.2', 'ipsrc': u'Static', 'gateway': u'0.0.0.0', 'vlanid': 11}}
mid05tor12cn02: BMC IP: 172.11.139.2
mid05tor12cn02: BMC Netmask: 16
mid05tor12cn02: BMC Gateway: 0.0.0.0 (default: 172.11.253.27)
mid05tor12cn02: BMC IP Source: Static
mid05tor12cn02: BMC VLAN ID: 11
mid05tor12cn02: BMC NTP Servers: []

### CODE REVIEW

**Comment by zet809** [comment]
The test for `rspconfig sshcfg:`

**Comment by gurevichmark** [comment]
Recommend that these 2 messages are only displayed when `-V` or debugmode is on:
```
Prepared /root/.ssh/./copy.sh file done
ip: 172.11.139.2, name: root, ps: 0penBmc
```

File: xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py [hunk 1]
```diff
@@ -115,18 +177,96 @@ def _set_netinfo(self, ip, netmask, gateway, vlan=False, **kw):
...
+                continue
+            dev,match,netid = k.partition("/ipv4/")
+            if netid:
+                if v["Origin"].find("LinkLocal") >= 0 or v["Address"].startswith("169.254"):
+                    self.callback.info("%s: Found LinkLocal address %s for interface %s, Ignoring..." % (node, v["Address"], dev))
+                    continue
+                nicid = dev.split('/')[-1]
+                if not netinfo.has_key(nicid):
+                    netinfo[nicid] = {}
+                if netinfo[nicid].has_key("ip"):
```
Comment (xuweibj) [inline xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py]: It seems that "has_key()" is not supported in Python3, I think use ``if 'ip' in netinfo[nicid]`` is better.

File: xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py [hunk 2]
```diff
@@ -115,18 +177,96 @@ def _set_netinfo(self, ip, netmask, gateway, vlan=False, **kw):
...
+                                   debugmode=self.debugmode, verbose=self.verbose)
+        try:
+            obmc.login()
+            data = obmc.get_netinfo()
+        except (SelfServerException, SelfClientException) as e:
+            self.callback.info('%s: %s' % (node, e.message))
+            return
+        netinfo = {}
+        for k, v in data.items():
+            if k.find("network/config") >= 0:
```
Comment (xuweibj) [inline xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py]: If just judge 'network/config' whether in string 'k', I think could use ``if 'network/config' in k:``.

File: xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py [hunk 3]
```diff
@@ -115,18 +177,96 @@ def _set_netinfo(self, ip, netmask, gateway, vlan=False, **kw):
...
+                if v.has_key("HostName"):
+                    netinfo["hostname"] = v["HostName"]
+                if v.has_key("DefaultGateway"):
+                    netinfo["defaultgateway"] = v["DefaultGateway"]
+                continue
+            dev,match,netid = k.partition("/ipv4/")
+            if netid:
+                if v["Origin"].find("LinkLocal") >= 0 or v["Address"].startswith("169.254"):
+                    self.callback.info("%s: Found LinkLocal address %s for interface %s, Ignoring..." % (node, v["Address"], dev))
```
Comment (xuweibj) [inline xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py]: Is this debug msg? Or normal msg?

File: xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py [hunk 4]
```diff
@@ -115,18 +177,96 @@ def _set_netinfo(self, ip, netmask, gateway, vlan=False, **kw):
             result = "set net(%s, %s, %s) for eth0" % (ip, netmask, gateway)
         return self.callback.info("set_netinfo %s" % result)
 
-    def _get_netinfo(self, ip=False, ipsrc=False, netmask=False, gateway=False, vlan=False, hostname=False, **kw):
-        result = ''
+    def _get_and_parse_netinfo(self, **kw):
```
Comment (xuweibj) [inline xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py]: I think we could put this content of function to file "openbmc_client.py".

**Comment by zet809** [comment]
hi, @gurevichmark and @xuweibj , modified based on your comments, pls help to review again, thx a lot!

File: xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py [hunk 5]
```diff
@@ -35,25 +43,88 @@ def dump_download(self, id, **kw):
...
+
 
     def set_ipdhcp(self, **kw):
-        return self.callback.info("set_ipdhcp")
+        node = kw['node']
+        obmc = openbmc.OpenBMCRest(name=node, nodeinfo=kw['nodeinfo'], messager=self.callback,
+                                   debugmode=self.debugmode, verbose=self.verbose)
+        try:
+            obmc.login()
+            obmc.set_ipdhcp(key)
```
Comment (xuweibj) [inline xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py]: What is the 'key'?

File: xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py [hunk 6]
```diff
@@ -35,25 +43,88 @@ def dump_download(self, id, **kw):
...
+        except (SelfServerException, SelfClientException) as e:
+            self.callback.info("%s: %s" % (node, e.message))
```
Comment (xuweibj) [inline xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py]: I think if get error here, should return.

**Comment by zet809** [comment]
Hi, @xuweibj , modified, pls review, thx!

**Comment by xuweibj** [comment]
Looks good to me. Agree to merge. Thanks @zet809 .

**Pull request merged**